wrapper/src/main/java/software/amazon/jdbc/DriverConnectionProvider.java

```diff
@@ -40,28 +42,29 @@ public DriverConnectionProvider(final java.sql.Driver driver) {
   /**
    * Called once per connection that needs to be created.
    *
-   * @param protocol The connection protocol (example "jdbc:mysql://")
+   * @param databaseDialect The connection DatabaseDialect
    * @param hostSpec The HostSpec containing the host-port information for the host to connect to
    * @param props The Properties to use for the connection
    * @return {@link Connection} resulting from the given connection information
    * @throws SQLException if an error occurs
    */
   @Override
   public Connection connect(
-      final @NonNull String protocol,
+      final @NonNull DatabaseDialect databaseDialect,
       final @NonNull HostSpec hostSpec,
       final @NonNull Properties props)
       throws SQLException {
     final String databaseName =
         PropertyDefinition.DATABASE.getString(props) != null
             ? PropertyDefinition.DATABASE.getString(props)
             : "";
-    final StringBuilder urlBuilder = new StringBuilder();
-    urlBuilder.append(protocol).append(hostSpec.getUrl()).append(databaseName);
+    final StringBuilder urlBuilder = new StringBuilder(databaseDialect.getURLScheme());
+    urlBuilder.append(hostSpec.getUrl()).append(databaseName);
 
     // In the case where we are connecting to MySQL using MariaDB driver,
     // we need to append "?permitMysqlScheme" to the connection URL
-    if (protocol.startsWith("jdbc:mysql:")
+    // TODO: this is a bit ugly
+    if (databaseDialect instanceof MariaDBDialect
```

Comment:
This should be `MySQLDialect`. It is for the case where we are connecting to a MySQL Engine with a MariaDB driver, and we are using the `"jdbc:mysql:"` protocol.

The `permitMysqlScheme` parameter is not required if we are using the `jdbc:mariadb` protocol directly

```suggestion
    if (databaseDialect instanceof MySQLDialect
```

Reply:
If we use jdbc:mysql we will load the mysql driver, no?

Reply:
This is for the scenario where only the mariadb driver is on the classpath. Similar to our AWS MySQL JDBC driver, the MariaDB driver supports the jdbc:mysql protocol too.

Reply:
I wasn't aware of that. However, I don't think we can make that presumption. That is I don't think the driver can assume that there is only one driver on the classpath.

Reply:
While I agree with you, this if clause isn't assuming there is only one driver on the classpath, it is to address the specific case where mariadb driver is used for the mysql protocol.

The other cases do not need this extra parameter `permitMysqlScheme`.

| Protocol     | Driver  | Requires permitMysqlScheme |
|--------------|---------|----------------------------|
| jdbc:mysql   | mysql   | false                      |
| jdbc:mysql   | mariadb | true                       |
| jdbc:mariadb | mariadb | false                      |